test/Feature/RpcSpec.hs

```diff
@@ -274,21 +274,102 @@ spec actualPgVersion =
           { matchHeaders = [matchContentTypeJson] }
 
     context "proc argument types" $ do
-      it "accepts a variety of arguments" $
```

Comment:
This test seems to be completely broken in it's previous form. @steve-chavez can you try to reproduce the following?

On current master change e.g. `"double": 3.1` to `"double": "x.1"` for this test. I would expect an error along the lines of "invalid input syntax for double precision". But for me, the test still passes.

When I then change the `varied_arguments` function to return the json object (see `schema.sql`) with all argument values - it correctly throws that input syntax error.

I was not able to reproduce that with bare SQL, but apparently we are calling the RPC in a way where that wouldn't throw.

In this state this test is basically useless, because you could pass anything and wouldn't notice whether it's cast correctly - that's why I changed the whole test to return the arguments in a json object as well.

Reply:
> In this state this test is basically useless, because you could pass anything and wouldn't notice whether it's cast correctly - that's why I changed the whole test to return the arguments in a json object as well.

@wolfgangwalther You're right. Looks like with the old test, the parameters are ignored if not used in the body. I tried this:

```

CREATE FUNCTION varied_arguments(
  double double precision,
  "varchar" character varying,
  "boolean" boolean,
  date date,
  money money,
  enum enum_menagerie_type,
  "integer" integer default 42,
  json json default '{}',
  jsonb jsonb default '{}'
) RETURNS text
AS $_$
  SELECT $1::text;
$_$ LANGUAGE sql;

http POST "l:3000/rpc/varied_arguments_plpgsql" <<JSON
{
              "double": "x.1",
              "varchar": "hello",
              "boolean": true,
              "date": "20190101",
              "money": 0,
              "enum": "foo",
              "integer": 43,
              "json": {"some key": "some value"},
              "jsonb": {"another key": [1, 2, "3"]}
            }
JSON

HTTP/1.1 400 Bad Request
{
    "code": "22P02",
    "details": null,
    "hint": null,
    "message": "invalid input syntax for type double precision: \"x.1\""
}
```
If the param is used then the syntax error happens.

Check this out, with PLPGSQL(instead of SQL) the function will fail despite the body:

```
CREATE FUNCTION varied_arguments_plpgsql(
  double double precision,
  "varchar" character varying,
  "boolean" boolean,
  date date,
  money money,
  enum enum_menagerie_type,
  "integer" integer default 42,
  json json default '{}',
  jsonb jsonb default '{}'
) RETURNS text
AS $_$ begin
  return 'Hi'::text;
end; $_$ LANGUAGE plpgsql;

http POST "l:3000/rpc/varied_arguments_plpgsql" <<JSON
{
              "double": "x.1",
              "varchar": "hello",
              "boolean": true,
              "date": "20190101",
              "money": 0,
              "enum": "foo",
              "integer": 43,
              "json": {"some key": "some value"},
              "jsonb": {"another key": [1, 2, "3"]}
            }
JSON

HTTP/1.1 400 Bad Request

{
    "code": "22P02",
    "details": null,
    "hint": null,
    "message": "invalid input syntax for type double precision: \"x.1\""
}
```

I was unable to reproduce with plain SQL though:

```
CREATE FUNCTION hey(
  par float8
) RETURNS text
AS $_$
  SELECT 'Hi'::text;
$_$ LANGUAGE sql;

select hey('x.1');
ERROR:  22P02: invalid input syntax for type double precision: "x.1"
LINE 1: select hey('x.1');
```

So the test change is good. My guess is that it has to be the way we wrap the function in our queries.

Reply:
Ok, that makes a lot of sense. I think the `SQL` functions / statements, are actually parsed at definition time - so postgres knows it won't need the arguments and will not even evaluate them (lazy evaluation). `plpgsql` functions on the other hand are parsed at execution time, so the arguments have to be cast before the functions is executed and the error occurs.

Reproducing this in plain sql queries did not work, I guess, because we were both using constants as arguments to the function - and I guess constants are evaluated and cast even before all of that, when parsing the `select hey('x.1')` query.